### PR TITLE
Prototype pollution fix - checking magic attributes - grpc-js

### PR DESCRIPTION
**Pull request by d3v53c**
### 📊 Metadata *

grpc is vulnerable to Prototype Pollution. This package allowing for modification of prototype behavior, which may result in Information Disclosure/DoS/RCE.

### ⚙️ Description *

Prototype Pollution refers to the ability to inject properties into existing JavaScript language construct prototypes, such as objects.
JavaScript allows all Object attributes to be altered, including their magical attributes such as proto, constructor and prototype.
An attacker manipulates these attributes to overwrite, or pollute, a JavaScript application object prototype of the base object by injecting other values.
Properties on the Object.prototype are then inherited by all the JavaScript objects through the prototype chain.

### 💻 Technical Description *

Fixed by avoiding setting magical attributes. The bug is fixed by validating the input strArray to check for prototypes. It is implemented by a simple validation to check for prototype keywords (proto, constructor and prototype), where if it exists, the function returns the object without modifying it, thus fixing the Prototype Pollution Vulnerability.

### 🐛 Proof of Concept (PoC) *

Create the following PoC file:

`// poc.js`
`var grpc =require('grpc')`
`grpc.loadPackageDefinition({'constructor.prototype.polluted': "Yes! Its Polluted"});`
`console.log({}.polluted)`

Execute the following commands in another terminal:

`npm i grpc # Install affected module`
`node poc.js #  Run the PoC`

Check the Output:

[Function: ServiceClient] { service: 'Yes! Its Polluted' }`

### 🔥 Proof of Fix (PoF) *

Before:
image

After:
image

### 👍 User Acceptance Testing (UAT)

After fix functionality is unaffected.

grpc-proto-test-1

grpc-proto-test-2
grpc-proto-test-3
grpc-proto-test-4
grpc-proto-test-5
grpc-proto-test-6
grpc-proto-test-7

### CODE REVIEW

File: packages/grpc-js/src/make-client.ts [hunk 1]
```diff
@@ -155,7 +164,7 @@ export function makeClientConstructor(
     ServiceClientImpl.prototype[name] = methodFunc;
     // Associate all provided attributes with the method
     Object.assign(ServiceClientImpl.prototype[name], attrs);
-    if (attrs.originalName && attrs.originalName !== '__proto__') {
+    if (attrs.originalName && isPrototypePolluted(attrs.originalName)) {
```
Comment (Mik317) [inline packages/grpc-js/src/make-client.ts]: 2' condition sentence was `attrs.originalName !== '__proto__'` , so the correct version should be `!isPrototypePolluted(attrs.originalName)`

**Review by Mik317 — CHANGES_REQUESTED** [review]
Hi @d3v53c :smile:,
thanks for the fix and it looks good to me.
I think there's only a simple mistake in the `if` sentence which makes still possible prototype pollution`, and should be replaced with a `!` value of the same.

Cheers,
Mik

File: packages/grpc-js/src/make-client.ts (hunk 1 above)
Comment (d3v53c) [reply packages/grpc-js/src/make-client.ts]: Thank you @Mik317, I've fixed the mistake. Appreciate the support.

Cheers

**Comment by d3v53c** [comment]
@Mik317 , Updated according to the requested changes.

Cheers

**Comment by d3v53c** [comment]
Closing this issue, and opening a new one #3 .

**Comment by huntr-helper** [comment]
Sorry d3v53c, we enjoyed reviewing your fix but it has not been selected this time. If this bounty has not been closed, please feel free to try again with a new pull request!

We appreciate your effort and look forward to reviewing more of your fixes in the future! 🔨😎

[Come join us on Discord](https://huntr.dev/discord)

**Pull request closed without merging**